drivers/flash/nrf_qspi_nor.c

```diff
@@ -262,13 +280,58 @@ static inline void qspi_wait_for_completion(const struct device *dev,
 	struct qspi_nor_data *dev_data = get_dev_data(dev);
 
 	if (res == NRFX_SUCCESS) {
+#ifdef CONFIG_MULTITHREADING
 		k_sem_take(&dev_data->sync, K_FOREVER);
+#else /* CONFIG_MULTITHREADING */
+
+/* We need to ensure that the interrupt that will set `ready` does not
+ * occur between the point we check `ready` and the point where we
+ * enter CPU idle.  There are two alternatives: k_cpu_idle() and
+ * k_cpu_atomic_idle().
+ *
+ * k_cpu_idle() (apparently) returns with interrupts unlocked even if
+ * it entered with them locked.
+ *
+ * k_cpu_atomic_idle(key) is to be called with interrupts locked to
+ * state key, and restores that state before it returns (for most
+ * architectures).
+ *
+ * So, two alternatives that seem to make sense:
```

Comment:
After reviewing alternatives there are two ways that should be safe.

I prefer the first, which uses `k_cpu_atomic_idle()`, because (ignoring incomplete documentation and a bad example) it makes the most sense relative to preserving the `key`.

However it's been suggested that `k_cpu_atomic_idle()` on ARM should be avoided because it hasn't been reworked like `k_cpu_idle()` has.  Thus the second approach.

Both work.  Which one do people want, or something else?

Reply:
I'm now strongly against the one that uses `k_cpu_idle()` because the documentation for that function has:
```
 * @note In some architectures, before returning, the function unmasks interrupts
 * unconditionally.
```
which means its behavior is not predictable.  While this particular code is only ever going to run on Nordic (ARM), there's no promise that ARM will always behave the way it does today, and the code fragment can't be copied into another architecture and still be expected to work.

Reply:
I have compiled LittleFS sample with the changes (not using `k_cpu_idle`) and with MULTITHREADING disabled and they seem to work fine.
I agree with @pabigot  argumentation that we should avoid using `k_cpu_idle`.
Shouldn't the `k_cpu_idle` be fixed to have consistent behavior across all architectures?

Reply:
Updated to leave only the `k_cpu_atomic_idle` solution.